juju/terraform-provider-juju
# [JUJU-2717] manage ssh keys

## Description

Fix #145 by adding support for SSH keys. 

There is a limitation in the way Juju manages SSH keys. For design reasons not all the SSH keys of a model can be removed. I have reported this in [LP2008087](https://bugs.launchpad.net/juju/+bug/2008087). To work around this limitation, I have added a dummy solution that does not deletes a key if it's the last available one and displays a warning message. This limitation directly impacts the user experience and may eventually lead to weird scenarios. Unfortunately, this is the only solution we can have until a decision is done on the Juju side.

## Type of change

Please mark if proceeds.


## Environment

- Juju controller version: 2.9.38
- Terraform version: 1.3

## QA steps

This is a simplistic test to see if we can add/remove SSH keys to Juju.

Apply  the following plan.
```tf
provider "juju" {}

resource "juju_model" "development" {
  name = "development"
}

resource "juju_ssh_key" "mykey" {
   model   = juju_model.development.name
   payload = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQC1I8QDP79MaHEIAlfh933zqcE8LyUt9doytF3YySBUDWippk8MAaKAJJtNb+Qsi+Kx/RsSY02VxMy9xRTp9d/Vr+U5BctKqhqf3ZkJdTIcy+z4hYpFS8A4bECJFHOnKIekIHD9glHkqzS5Vm6E4g/KMNkKylHKlDXOafhNZAiJ1ynxaZIuedrceFJNC47HnocQEtusPKpR09HGXXYhKMEubgF5tsTO4ks6pplMPvbdjxYcVOg4Wv0N/LJ4ffAucG9edMcKOTnKqZycqqZPE6KsTpSZMJi2Kl3mBrJE7JbR1YMlNwG6NlUIdIqVoTLZgLsTEkHqWi6OExykbVTqFuoWJJY3BmRAcP9H3FdLYbqcajfWshwvPM2AmYb8V3zBvzEKL1rpvG26fd3kGhk3Vu07qAUhHLMi3P0McEky4cLiEWgI7UyHFLI2yMRZgz23UUtxhRSkvCJagRlVG/s4yoylzBQJir8G3qmb36WjBXxpqAGHfLxw05EQI1JGV3ReYOs= bbbb@somewhere"
}
```

```bash
terraform init
terraform apply
```

Juju shows the new key
```bash
juju ssh-keys -m development
Keys used in model: development
03:62:17:d0:72:87:37:3d:f9:4f:f3:74:c6:c3:26:49 (bbbb@somewhere)
```

Add another key

```tf
provider "juju" {}

resource "juju_model" "development" {
  name = "development"
}

resource "juju_ssh_key" "mykey" {
   model   = juju_model.development.name
   payload = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQC1I8QDP79MaHEIAlfh933zqcE8LyUt9doytF3YySBUDWippk8MAaKAJJtNb+Qsi+Kx/RsSY02VxMy9xRTp9d/Vr+U5BctKqhqf3ZkJdTIcy+z4hYpFS8A4bECJFHOnKIekIHD9glHkqzS5Vm6E4g/KMNkKylHKlDXOafhNZAiJ1ynxaZIuedrceFJNC47HnocQEtusPKpR09HGXXYhKMEubgF5tsTO4ks6pplMPvbdjxYcVOg4Wv0N/LJ4ffAucG9edMcKOTnKqZycqqZPE6KsTpSZMJi2Kl3mBrJE7JbR1YMlNwG6NlUIdIqVoTLZgLsTEkHqWi6OExykbVTqFuoWJJY3BmRAcP9H3FdLYbqcajfWshwvPM2AmYb8V3zBvzEKL1rpvG26fd3kGhk3Vu07qAUhHLMi3P0McEky4cLiEWgI7UyHFLI2yMRZgz23UUtxhRSkvCJagRlVG/s4yoylzBQJir8G3qmb36WjBXxpqAGHfLxw05EQI1JGV3ReYOs= bbbb@somewhere"
}

resource "juju_ssh_key" "myotherkey" {
   model   = juju_model.development.name
   payload = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQC1I8QDP79MaHEIAlfh933zqcE8LyUt9doytF3YySBUDWippk8MAaKAJJtNb+Qsi+Kx/RsSY02VxMy9xRTp9d/Vr+U5BctKqhqf3ZkJdTIcy+z4hYpFS8A4bECJFHOnKIekIHD9glHkqzS5Vm6E4g/KMNkKylHKlDXOafhNZAiJ1ynxaZIuedrceFJNC47HnocQEtusPKpR09HGXXYhKMEubgF5tsTO4ks6pplMPvbdjxYcVOg4Wv0N/LJ4ffAucG9edMcKOTnKqZycqqZPE6KsTpSZMJi2Kl3mBrJE7JbR1YMlNwG6NlUIdIqVoTLZgLsTEkHqWi6OExykbVTqFuoWJJY3BmRAcP9H3FdLYbqcajfWshwvPM2AmYb8V3zBvzEKL1rpvG26fd3kGhk3Vu07qAUhHLMi3P0McEky4cLiEWgI7UyHFLI2yMRZgz23UUtxhRSkvCJAgRlVG/s4yoylzBQJir8G3qmb36WjBXXpqAGHfLxw05EQI1JGV3ReYOs= cccc@somewhere"
}
```

Apply and check there is a new key
```bash
juju_29 ssh-keys -m development
Keys used in model: development
ba:fd:cc:51:d3:6b:18:a6:4a:34:19:81:a6:d8:95:82 (cccc@somewhere)
03:62:17:d0:72:87:37:3d:f9:4f:f3:74:c6:c3:26:49 (bbbb@somewhere)
```

Make any change in the last added key. This should force a recreation of the resource. Something like this should be displayed
```bash
 # juju_ssh_key.myotherkey must be replaced
-/+ resource "juju_ssh_key" "myotherkey" {
      ~ id      = "sshkey:development:cccc@somewhere" -> (known after apply)
      ~ payload = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQC1I8QDP79MaHEIAlfh933zqcE8LyUt9doytF3YySBUDWippk8MAaKAJJtNb+Qsi+Kx/RsSY02VxMy9xRTp9d/Vr+U5BctKqhqf3ZkJdTIcy+z4hYpFS8A4bECJFHOnKIekIHD9glHkqzS5Vm6E4g/KMNkKylHKlDXOafhNZAiJ1ynxaZIuedrceFJNC47HnocQEtusPKpR09HGXXYhKMEubgF5tsTO4ks6pplMPvbdjxYcVOg4Wv0N/LJ4ffAucG9edMcKOTnKqZycqqZPE6KsTpSZMJi2Kl3mBrJE7JbR1YMlNwG6NlUIdIqVoTLZgLsTEkHqWi6OExykbVTqFuoWJJY3BmRAcP9H3FdLYbqcajfWshwvPM2AmYb8V3zBvzEKL1rpvG26fd3kGhk3Vu07qAUhHLMi3P0McEky4cLiEWgI7UyHFLI2yMRZgz23UUtxhRSkvCJAgRlVG/s4yoylzBQJir8G3qmb36WjBXXpqAGHfLxw05EQI1JGV3ReYOs= cccc@somewhere" -> "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQC1I8QDP79MaHEIAlfh933zqcE8LyUt9doytF3YySBUDWippk8MAaKAJJtNb+Qsi+Kx/RsSY02VxMy9xRTp9d/Vr+U5BctKqhqf3ZkJdTIcy+z4hYpFS8A4bECJFHOnKIekIHD9glHkqzS5Vm6E4g/KMNkKylHKlDXOafhNZAiJ1ynxaZIuedrceFJNC47HnocQEtusPKpR09HGXXYhKMEubgF5tsTO4ks6pplMPvbdjxYcVOg4Wv0N/LJ4ffAucG9edMcKOTnKqZycqqZPE6KsTpSZMJi2Kl3mBrJE7JbR1YMlNwG6NlUIdIqVoTLZgLsTEkHqWi6OExykbVTqFuoWJJY3BmRAcP9H3FdLYbqcajfWshwvPM2AmYb8V3zBvzEKL1rpvG26fd3kGhk3Vu07qAUhHLMi3P0McEky4cLiEWgI7UyHFLI2yMRZgz23UUtxhRSkvCJAgRlVG/s4yoylzBQJir8G3qmb36WjBXwpqAGHvLxw05EQI1JGV3ReYOs= cccc@somewhere" # forces replacement
        # (1 unchanged attribute hidden)
    }
```

Now we remove the last key from the plan, apply, and check that the `cccc@somewhere` key disappears.

```bash
juju_29 ssh-keys -m development
Keys used in model: development
03:62:17:d0:72:87:37:3d:f9:4f:f3:74:c6:c3:26:49 (bbbb@somewhere)
```

# Additional notes

Ongoing bug reported [here](https://bugs.launchpad.net/juju/+bug/2008087)
